src/server/pfs/server/driver.go

```diff
@@ -3254,53 +3267,38 @@ func (d *driver) copyFile(pachClient *client.APIClient, src *pfs.File, dst *pfs.
 
 func (d *driver) getTreeForCommit(txnCtx *txnenv.TransactionContext, commit *pfs.Commit) (hashtree.HashTree, error) {
 	if commit == nil || commit.ID == "" {
-		t, err := hashtree.NewDBHashTree(d.storageRoot)
-		if err != nil {
-			return nil, err
-		}
-		return t, nil
+		return d.treeCache.GetOrAdd("nil", func() (hashtree.HashTree, error) {
+			return hashtree.NewDBHashTree(d.storageRoot)
+		})
 	}
 
-	tree, ok := d.treeCache.Get(commit.ID)
-	if ok {
-		h, ok := tree.(hashtree.HashTree)
-		if ok {
-			return h, nil
+	return d.treeCache.GetOrAdd(commit.ID, func() (hashtree.HashTree, error) {
```

Comment:
Adding a tree to the cache now must be done atomically - the old system was a bit racy and could result in a hashtree leak.  Note that this may have some performance implications based on how long we expect this operation to take.